### ISSUE TITLE
Feature Request - Support for Alignment Matrix Control

### ISSUE THREAD

**Issue by nick6352683** [AdvancedCustomFields/acf#309]
Is it possible for ACF blocks to support for the brand new Alignment Matrix Control, which was added to the Cover Block? (https://github.com/WordPress/gutenberg/tree/master/packages/components/src/alignment-matrix-control)

If not, can I request this as a future feature?

Thanks,
Nick.

**Comment by elliotcondon** [comment]
Hi @nick6352683 

Thanks for the feature request. The new `<AlignmentMatrixControl />` component looks great. I just played around with the Cover block in Gutenberg 8.2 and can see how this would be a very useful piece in the block building puzzle.

I'll have a think about how we could add support for custom elements such as this one within the "Block Controls".

My first thought is to use a JS filter/action and allow developers full access to customize the components shown here, but will also consider a more "PHP" friendly solution too.

I also wonder if it is possible to wrap each ACF Block in a "Higher-Order Component" and completely customize the way ACF Blocks are rendered (to add in this Component + logic to set the alignment).

Thanks
Elliot

**Comment by nick6352683** [comment]
That's great Elliot, thanks. I am sure, and I'm talking on my behalf anyway that a more "PHP" friendly solution would be greatly appreciated.

**Comment by elliotcondon** [comment]
Let's explore this a little further...

a) Would the new `<AlignmentMatrixControl />` control be used in-place of the standard "align" control? If so, perhaps we could add a new "support" setting that simply uses this Component instead of the default one.

b) Opposite to the above, would there be a situation where a developer would use both the default "align" control and also the new `<AlignmentMatrixControl />` one? My gut feeling is no, but it's worth asking.

c) Are there any other "Block Controls" that could be added / swapped with other PHP settings?

**Comment by nick6352683** [comment]
Actually, in my situation is exactly like b). With CSS settings I use the alignwide class to set the background as fullwidth, but keeping the content the same. I use the alignfull class to extend both the background and content at the edge of the browser window. 

That said, the alignment matrix control is where the content of the container will be aligned. Basically I like exactly how the cover blocks is working.

Let me know if you want a video demo to see how I setup the allignwide and alignfull classes in my themes.

**Comment by elliotcondon** [comment]
Aah yes, that makes sense. The "default align control" is used to set the block's width/alignment, whereas the "matrix align control" is used to set the position of inner content. 

Having both controls visible next to each other makes sense.

Looking at the [acf_register_block_type](https://www.advancedcustomfields.com/resources/acf_register_block_type/) "supports" setting, it would make sense to follow a similar pattern to the existing "align" property. What do you think about adding a new support property named "align_matrix" with the ability to set this to `false`, `true`, or perhaps even a string of allowed positions (if the Component allows for this)?

**Comment by nick6352683** [comment]
"Looking at the acf_register_block_type "supports" setting, it would make sense to follow a similar pattern to the existing "align" property."

That is exactly what I was hoping for... we are on the same page on this.

Thanks Elliot.

**Comment by JiveDig** [comment]
I think an array would be nice. I can foresee use-cases where you'd only want top/center/bottom/left/right and not top-right an corners. Or just horizontal or vertical. It would be cool to use the same UI for alignment even when you only need a few.

Right now we're using a button group set to left/center/right and another for top/middle/bottom.

**Comment by elliotcondon** [comment]
Thanks @JiveDig. I've got this feature on my to-do list for beta4 and am really looking forward to playing with the new Alignment Matrix button!

**Comment by elliotcondon** [comment]
@nick6352683 @JiveDig Good news. I've got an initial working solution for the Alignment Matrix Control 🎉.

Everything seems pretty straight forward, except for the naming. The WordPress Cover block is using the naming `contentPosition`, but I believe `align_content` would work best for ACF. This way, when specifying ACF Blocks settings, the naming would be uniform between "align" and "align_content" like so:
```
acf_register_block_type( array(
    'name'				=> 'test_align_matrix',
    'title'				=> 'Test Align Matrix',
    ...
    'supports'			=> array(
        'align' => true,
        'align_content' => true,
    )
));
```

Thoughts?

**Comment by nick6352683** [comment]
That sounds amazing, and very easy for us to implement. I'm guessing this would be included in the next beta version (5.9 beta4)...

Once again, thanks a million Elliot !

**Comment by elliotcondon** [comment]
@nick6352683 Thanks mate! I've just pushed this update to the `feature/release-5-9` branch which includes support for the Alignment Matrix Control component.

Documentation doesn't exist yet on the ACF site, so here's how to use it:
```
acf_register_block_type( array(
    'name'			=> 'test_align_matrix',
    'title'			=> 'Test Align Matrix',
    // ...
    'align'			=> 'wide',
    'align_content'		=> 'center right',
    'supports'			=> array(
        'align' => true,
        'align_content' => true,
    ),
    'render_callback'	=> function( $block ){
        echo '<div class="test_align_matrix">';
            echo '<p>' . $block['align'] . ' ' . $block['align_content'] . '</p>';
        echo '</div>';
    }
));
```

notes:
1. Register support for "align_content" to enable the toolbar component.
2. Provide an optional default value (similar to "align"). Remove this for a default "center center" value.

**Comment by elliotcondon** [comment]
... and yes, this will be included in beta4

**Comment by nick6352683** [comment]
Amazing, thanks 1

**Comment by JiveDig** [comment]
Ah interesting. So the matrix is only a per-block thing like how align is now? 

I was imagining it as a new field type tbh, so you can use that field type to control different alignment of various elements in your block. 

Maybe that’s not possible or not the scope, and maybe a per-block setting is all I’d need, but figured I’d mention it.

**Comment by JiveDig** [comment]
For reference, I was thinking along the lines of Kirki’s dimensions control, where you can use it however you want: https://kirki.org/docs/controls/dimensions/

**Comment by elliotcondon** [comment]
@JiveDig The dimensions field looks cool, but yeah this is something completely different 👍

**Comment by CreativeDive** [comment]
@elliotcondon I've tested your new matrix control feature for ACF blocks. I think it is a little bit different to the default matrix control. The ACF matrix control provides always a default value. Also if I add a new block. There is the "center center" position set. But by default, the matrix control should not have a default position. Like the cover block.

It's possible to provide a solution to disable the default value, like `'align_content'	=> false`?

**Comment by elliotcondon** [comment]
Hi @CreativeDive. I've just tested the core Cover block again and can confirm that the default value shown in the UI is "center center". It may be that the JS attributes allow for an empty value, which would be confusing, as there is no way to "un select" a selected position. The empty attribute value should always be treated as "center center".

Here's a quick recording to demonstrate the default position of a Cover block.

![ScreenFlow](https://user-images.githubusercontent.com/2296425/88239903-0144cb80-ccc9-11ea-9b48-b5dd40d3c167.gif)

**Comment by CreativeDive** [comment]
@elliotcondon Yes I can confirm, this missing default value on block creation was fixed with the latest gutenberg release. So I think everything is fine, how this ACF feature works :-)

**Comment by elliotcondon** [comment]
Perfect!

**Comment by leroy** [comment]
@elliotcondon is this available in the latest 5.9 release?

**Comment by djpane** [comment]
I'm puzzled. I wanted to test the matrix control but all I get is the vertical alignment control. Even with the version 5.9.0 release note's example found here: https://www.advancedcustomfields.com/blog/acf-5-9-0-release-inner-blocks-ui-improvements-and-exciting-features/

`acf_register_block_type( array(
    'name'				=> 'test_align_matrix',
    'title'				=> 'Test Align Matrix',
    'align_content'		=> 'top left', // Specifies the default attribute value.
    'supports'			=> array(
        'align_content' => true // Enables the feature.
    ),
    'render_callback'	=> function( $block ){
        echo '<div class="is-position-' . esc_attr( $block['align_content'] ) . '">';
            echo '<p>' . esc_html( $block['align_content'] ) . '</p>';
        echo '</div>';
    }
));`

I have ACF 5.9.9 and WP 5.8.
